### DISCUSSION TITLE
VNA calibration and error correction references.

### DISCUSSION THREAD

**Discussion by trcwm** [scott-guthridge/libvna#3] [Q&A] [answered]
Dear Mr. Guthridge,

I am playing around with custom NanoVNA firmware and would like to know more about calibration techniques, and I see your library has quite a few. The papers on the subject I've found (Rehnmark 1974 and the HP Journal Feb 1970) are rather short and specific. Do you know of any good references that are a little more general, perhaps some kind of handbook? 

Thank you for your time.

Regards,
Niels Moseley.

**Comment by scott-guthridge** [comment, marked as answer]
There are many papers on the topic in IEEE transactions on Microwave Theory and Techniques dating back over 60 years  (if you have access).  I've read most of them.  It's interesting to see how the topic has evolved over the years and how some of the old papers are still relevant today.

It sounds like you're looking more for a good introduction and overview rather than academic papers, though.  This [slide deck](https://pdfslide.net/documents/network-analyzer-error-models-and-calibration-rytting-this-paper-is-an-overview.html?page=1) is a pretty good overview of the "12 term", "8 term" and "16 term" models.  I would start with that.  

In the slide deck, I think Rytting could have explained a few details better, though.  For example, there is a matrix on page 37 with elements T1, T2, T3 and T4 describing a 4-port error box system in T parameters.  T parameters are, by definition, 2x2, so one might ask: how can T parameters be used to describe that 4-port system?  And how would it work if we generalize, to say, a 4-port VNA?  The answer (that the paper doesn't explicitly explain) is that the T1, T2, T3, and T4 matrices are not T parameter matrices themselves; rather, the T parameter matrix is the matrix formed out of them:

    T1 T2
    T3 T4

where the elements are, themselves, matrices.  In libvna, I call these four elements: Ts, Ti, Tx and Tm, respectively, indicating whether they multiply the actual S parameters (s), the identify matrix (i), both actual and measured (x) or just the measured parameters (m).  See the comments in src/vnacal_layout.h to compare.  If we generalize to a 4-port VNA, there would still be four of these element matrices, but their dimensions all become 4x4.   In the last section, Rytting makes the case that the 12-term parameters handle the same errors as the 8-term parameters, so there must be dependencies between the error terms.  That's true, but only if you have the "a" (reference) measurements to compensate for errors in the forward-reverse switch.

You can find many white papers from test and measurement equipment vendors if you do an Internet search for "VNA calibration".  I hope this helps.
